File: src/main/java/sort/algo/QuickSort.java
package sort.algo;

public class QuickSort implements SortAlgorithm {

    /**
     * Time:    O(n log (n)); avg case, O(n^2) worst case
     * Space:   O(log (n))
     * n is the size of the array
     */
    public void sort(int[] array) {
        quickSort(array, 0, array.length - 1);
    }

    private void quickSort(int[] array, int left, int right) {
        int index = partition(array, left, right);

        if (index == -1) return;
        if (left < index - 1) quickSort(array, left, index - 1);
        if (index < right) quickSort(array, index, right);
    }

    private int partition(int[] array, int left, int right) {
        if (array == null || array.length == 0) return -1;

        int pivot = array[(left + right) / 2];

        while (left <= right) {
            while (array[left] < pivot) left++;
            while (array[right] > pivot) right--;

            if (left <= right) {
                swap(array, left, right);
                left++;
                right--;
            }
        }

        return left;
    }

    private void swap(int[] array, int i1, int i2) {
        int tmp = array[i1];
        array[i1] = array[i2];
        array[i2] = tmp;
    }
}

File: src/main/python/sort/algo/quick_sort.py
def quicksort(values):
    def qsort(start, end):
        if start >= end:
            return

        pivot = values[(start + end) // 2]
        index = partition(start, end, pivot)
        qsort(start, index - 1)
        qsort(index, end)

    def partition(start, end, pivot):
        while start <= end:
            while values[start] < pivot:
                start += 1

            while values[end] > pivot:
                end -= 1

            if start <= end:
                swap(start, end)
                start += 1
                end -= 1

        return start

    def swap(first, second):
        temp = values[first]
        values[first] = values[second]
        values[second] = temp

    qsort(0, len(values) - 1)


if __name__ == '__main__':
    values1 = [9, 7, 16, 20, 13, 100, -5, 34, 2, 27]
    print(values1)
    quicksort(values1)
    print(values1)
